File: services/clients/node-pg/probe.mjs
// Client compatibility fixture: node-postgres.
//
// The same six checks as services/clients/psql/probe.sh, through a driver that
// is not libpq. That is the point of running both: psql answers "what does the
// protocol offer", and this answers "what does a driver that reimplements the
// protocol actually do with it".
//
// THE CREDENTIAL IS NOT AN ENVIRONMENT VARIABLE. node-postgres depends on
// `pgpass` (1.0.5 in the committed lock file, a transitive dependency of pg
// 8.16.3), which reads `process.env.PGPASSFILE` when no password is supplied.
// So this fixture passes no password at all and lets the driver find the file
// the entrypoint wrote -- the same file `bin/connect.sh exec` provides. That
// was read off the resolved dependency tree rather than assumed: a driver
// without pgpass support would need PGPASSWORD, and PGPASSWORD is banned from
// this model outright.
//
// Exit codes: 0 every check passed, 2 bad input, 6 a check failed.

import pg from "pg";

const TRANSPORT = required("APG_TRANSPORT");
const USER_A = required("APG_USER_A");
const USER_B = required("APG_USER_B");
const APPLICATION_NAME = `apg-client-node-pg-${TRANSPORT}`;

function required(name) {
  const value = process.env[name];
  if (!value) {
    console.error(`client-node-pg: ${name} is required`);
    process.exit(2);
  }
  return value;
}

function fail(label, expected, got) {
  console.error(`client-node-pg: ${label}: expected ${expected}, got ${got}`);
  process.exit(6);
}

function expect(label, expected, got) {
  // Compared with Object.is on strings, never on a truthy value. A count that
  // arrives as the string "0" is truthy in JavaScript, which is precisely the
  // way an isolation check passes while measuring nothing.
  if (String(got) !== String(expected)) {
    fail(label, expected, got);
  }
  console.log(`  ok    ${label}`);
}

if (!["pooled", "direct"].includes(TRANSPORT)) {
  console.error(`client-node-pg: APG_TRANSPORT must be pooled or direct, not ${TRANSPORT}`);
  process.exit(2);
}

// PGHOST, PGPORT, PGUSER and PGDATABASE come from the environment the way libpq
// reads them; pg honours all four. `application_name` is set here rather than
// left to the caller because it is what makes a pooled server connection
// attributable in pg_stat_activity.
const client = new pg.Client({ application_name: APPLICATION_NAME });

// Every statement below is parameterized. There is no string concatenation into
// SQL anywhere in this file, and a claim is set with set_config(..., true)
// rather than `SET LOCAL`, because SET takes no parameters -- which is the
// reason a driver that only offered SET would have to interpolate.
async function claim(userId) {
  await client.query("SELECT set_config('app.user_id', $1, true)", [userId]);
}

async function main() {
  await client.connect();
  console.log(
    `client-node-pg: ${process.env.PGDATABASE} over the ${TRANSPORT} transport as ${process.env.PGUSER}`,
  );

  // 1. The application name survives the transport. Read back from the server,
  //    not from the object that set it.
  const named = await client.query(
    "SELECT application_name FROM pg_stat_activity WHERE pid = pg_backend_pid()",
  );
  expect("application_name reaches the server", APPLICATION_NAME, named.rows[0].application_name);

  // 2 and 3. A write under each user's claim, each inside an explicit
  //          transaction. Explicit because the claim is transaction-local: in
  //          autocommit every statement is its own transaction, so the claim
  //          would be gone before the insert that depends on it.
  for (const [label, userId, title] of [
    ["user A", USER_A, "node-pg fixture note (A)"],
    ["user B", USER_B, "node-pg fixture note (B)"],
  ]) {
    await client.query("BEGIN");
    await claim(userId);
    const created = await client.query("SELECT (api.create_note($1)).owner_id AS owner_id", [
      title,
    ]);
    await client.query("COMMIT");
    expect(`${label} writes under its own claim`, userId, created.rows[0].owner_id);
  }

  // 4. Isolation as a pair of counts. "None of the other user's" alone would be
  //    satisfied by an empty table; "some of my own" alone would be satisfied by
  //    no policy at all.
  for (const [label, userId] of [
    ["user A", USER_A],
    ["user B", USER_B],
  ]) {
    await client.query("BEGIN");
    await claim(userId);
    const counted = await client.query(
      `SELECT count(*) FILTER (WHERE owner_id = $1::uuid) AS mine,
              count(*) FILTER (WHERE owner_id <> $1::uuid) AS theirs
         FROM api.notes`,
      [userId],
    );
    await client.query("COMMIT");
    const { mine, theirs } = counted.rows[0];
    if (Number(mine) < 1) fail(`${label} sees its own rows`, ">= 1", mine);
    expect(`${label} sees none of the other user's rows`, "0", theirs);
  }

  // 5. No claim, no rows. Run after the writes, so an empty table cannot be
  //    what makes it pass.
  await client.query("BEGIN");
  const unclaimed = await client.query("SELECT count(*) AS visible FROM api.notes");
  await client.query("COMMIT");
  expect("a transaction with no claim sees nothing", "0", unclaimed.rows[0].visible);

  // 6. The private schema is not addressable. Proved by attempting the read,
  //    not by asking the catalog: D103 measured has_table_privilege returning
  //    true for app.notes while the read is denied, because the boundary is the
  //    schema and the table grant is what makes the security-invoker views work.
  await client.query("BEGIN");
  let reachable = true;
  try {
    await client.query("SELECT count(*) FROM app.notes");
  } catch (error) {
    reachable = false;
    if (!/schema "?app"?/.test(error.message)) {
      fail("app is refused for the documented reason", "permission denied for schema app", error.message);
    }
  }
  await client.query("ROLLBACK");
  if (reachable) fail("app.notes is not readable", "permission denied", "rows returned");
  console.log("  ok    app.notes is not readable, whatever the table privilege says");

  await client.end();
  console.log("client-node-pg: every check passed");
}

main().catch((error) => {
  console.error(`client-node-pg: ${error.message}`);
  process.exit(6);
});

File: services/clients/psycopg/probe.py
"""Client compatibility fixture: Psycopg 3.

The same six checks as the psql and node-postgres fixtures, through the third
kind of client: a driver built on libpq. So `PGPASSFILE`, `PGAPPNAME` and the
rest are honoured by the C library rather than reimplemented, and what this
fixture actually tests is whether *its own* layer -- server-side binding,
transaction handling, and the pipeline it may use over the pooler -- behaves
through a transaction pooler.

**Every statement is parameterized, and Psycopg 3 binds server-side.** Psycopg 3
sends parameters over the extended query protocol rather than interpolating them
client-side the way Psycopg 2 did, which is what makes `%s` here a placeholder
rather than a format specifier. A claim is set with `set_config(..., true)`
rather than `SET LOCAL`, because `SET` takes no parameters.

**The transaction is explicit.** `autocommit` stays False and each block is a
`with connection.transaction()`, because the claim is transaction-local: under
autocommit every statement is its own transaction and the claim would be gone
before the insert that depends on it. That is the failure this fixture would see
first if the pooler's mode were changed to `session` and back.

Exit codes: 0 every check passed, 2 bad input, 6 a check failed.
"""

from __future__ import annotations

import os
import sys

import psycopg

TRANSPORTS = ("pooled", "direct")


def required(name: str) -> str:
    value = os.environ.get(name)
    if not value:
        print(f"client-psycopg: {name} is required", file=sys.stderr)
        raise SystemExit(2)
    return value


def fail(label: str, expected: object, got: object) -> None:
    print(f"client-psycopg: {label}: expected {expected}, got {got}", file=sys.stderr)
    raise SystemExit(6)


def expect(label: str, expected: object, got: object) -> None:
    if str(got) != str(expected):
        fail(label, expected, got)
    print(f"  ok    {label}")


def main() -> int:
    transport = required("APG_TRANSPORT")
    if transport not in TRANSPORTS:
        print(
            f"client-psycopg: APG_TRANSPORT must be pooled or direct, not {transport}",
            file=sys.stderr,
        )
        return 2

    user_a = required("APG_USER_A")
    user_b = required("APG_USER_B")
    application_name = f"apg-client-psycopg-{transport}"

    # No connection string is built here. libpq reads PGHOST, PGPORT, PGUSER,
    # PGDATABASE and PGPASSFILE from the environment, so there is nothing to
    # assemble and therefore nothing to assemble wrongly -- and no place a
    # credential could end up in a string.
    with psycopg.connect(application_name=application_name, autocommit=False) as connection:
        print(
            f"client-psycopg: {os.environ.get('PGDATABASE')} over the {transport} "
            f"transport as {os.environ.get('PGUSER')}"
        )

        # 1. The application name survives the transport, read back from the
        #    server rather than from the object that set it.
        with connection.transaction():
            row = connection.execute(
                "SELECT application_name FROM pg_stat_activity WHERE pid = pg_backend_pid()"
            ).fetchone()
        expect("application_name reaches the server", application_name, row[0])

        # 2 and 3. A write under each user's claim, inside an explicit
        #          transaction. create_note takes no owner parameter; it derives
        #          ownership from the claim, so the equality is a check that the
        #          derivation happened.
        for label, user_id, title in (
            ("user A", user_a, "psycopg fixture note (A)"),
            ("user B", user_b, "psycopg fixture note (B)"),
        ):
            with connection.transaction():
                connection.execute("SELECT set_config('app.user_id', %s, true)", (user_id,))
                created = connection.execute(
                    "SELECT (api.create_note(%s)).owner_id", (title,)
                ).fetchone()
            expect(f"{label} writes under its own claim", user_id, created[0])

        # 4. Isolation as a pair of counts. Either count alone can be satisfied
        #    by something other than a working policy.
        for label, user_id in (("user A", user_a), ("user B", user_b)):
            with connection.transaction():
                connection.execute("SELECT set_config('app.user_id', %s, true)", (user_id,))
                mine, theirs = connection.execute(
                    "SELECT count(*) FILTER (WHERE owner_id = %(id)s::uuid), "
                    "       count(*) FILTER (WHERE owner_id <> %(id)s::uuid) "
                    "  FROM api.notes",
                    {"id": user_id},
                ).fetchone()
            if mine < 1:
                fail(f"{label} sees its own rows", ">= 1", mine)
            expect(f"{label} sees none of the other user's rows", 0, theirs)

        # 5. No claim, no rows. Last, so the rows written above are what it has
        #    to fail to see.
        with connection.transaction():
            visible = connection.execute("SELECT count(*) FROM api.notes").fetchone()[0]
        expect("a transaction with no claim sees nothing", 0, visible)

        # 6. The private schema is not addressable, proved by attempting the
        #    read. D103: has_table_privilege reports true for app.notes because
        #    the role inherits that grant through `authenticated`, which is what
        #    makes the security-invoker views work; the schema is the boundary.
        reachable = True
        try:
            with connection.transaction():
                connection.execute("SELECT count(*) FROM app.notes")
        except psycopg.errors.InsufficientPrivilege as error:
            reachable = False
            if "schema" not in str(error):
                fail("app is refused for the documented reason", "schema app", str(error))
        if reachable:
            fail("app.notes is not readable", "permission denied", "rows returned")
        print("  ok    app.notes is not readable, whatever the table privilege says")

    print("client-psycopg: every check passed")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
